### ISSUE TITLE
no-internal-modules should ignore index.js imports (ESM)

### ISSUE THREAD

**Issue by nickluger** [import-js/eslint-plugin-import#2524]
For a directory `foo`
```
foo
  foo.js
  bar.js
  index.js
```
And using [ECMAScript modules (ESM)](https://nodejs.org/api/esm.html#modules-ecmascript-modules), Node.js module resolution is not a valid import syntax.
```
import { foo } from "./foo";
```
While these are valid, but being reported by `no-internal-modules`:

```
import { foo } from "./foo/index.js";
export * from "./foo/index.js";
```

`no-internal-modules` should be usable with ESM, though and ignore `/index.js` imports.

**Comment by JounQin** [comment]
When not just disable `no-internal-modules` rule in this case? And what should be reported with this rule enabled with ESM in your opinion then?

**Comment by ljharb** [comment]
Currently, this plugin only supports transpiled ESM, not native. If you use babel or tsc, this works fine.

When `resolve` supports the exports field and ESM resolution, we can also add an option to enable native ESM (which is strictly worse than transpiling in all envs, and so won’t be the recommended default)

**Comment by nickluger** [comment]
Ah, okay, thanks for the quick reply! 🙏

Yeah, but it's a bit of a PITA, as more and more parts of the ecosystem are moving [pretty aggressively](https://github.com/sindresorhus/meta/discussions/15) towards native ESM. With transpiling, I assume you mean transpiling to CJS in a Node environment? Cause, one can transpile to native ESM, too (`nodenext`-mode), which we do for internal libraries that are consumed by both environments.

Maybe I stirred up a hornet's nest here, but I don't really have an opinion on these things, would just be happy to deliver features, while still being able to use libraries…

> which is strictly worse than transpiling in all envs

In what way? I'd be great if you could just point me to an article or discussion where I could read upon that?

Thanks!

**Comment by ljharb** [comment]
That's not "more parts", that's just one single individual who unfortunately has an outsized reach. The vast majority of the ecosystem is sticking with what works for users, which is decidedly _not_ ESM-only - if you find any author prioritizing ideology over your actual use cases, I always suggest finding an alternative package with an author who is not hostile to your needs.

**Comment by nickluger** [comment]
Yeah, and this might be a bit off-topic (in this issue). But, when researching this stuff, you find articles from proponents for ESM, explaining why and how to move, or articles that explain why CJS and ESM don't get along.

I really valued your input on other issues, and especially because you have a bit different stance on that, I would like to know why you prefer `transpiling in all envs` and then how to deal with libraries?

**Comment by ljharb** [comment]
I prefer it because native ESM is slower and less featureful in every environment than CJS (bundled CJS in a browser). Libraries should be published as CJS (they can add ESM if they want, but there's no value to doing that), and then they can be used everywhere. ESM-only libraries still can't be used in a browser without a bundler, and native ESM in node doesn't yet have anywhere close to the featureset or ecosystem tooling that would be needed to compete with CJS.

Perhaps this will change one day, but that day is not anytime soon.

**Comment by nickluger** [comment]
Thank you for sharing your thinking on that!
